### ISSUE TITLE
Support white labeling OBA

### ISSUE THREAD

**Issue by aaronbrethorst** [OneBusAway/onebusaway-iphone#375]
Things that have to be done off the top of my head:
- [ ] Ensure all strings that reference the app name pull it from the app bundle
- [ ] Move all color and font references into a new class (tentatively named `OBAStyles`)
- [ ] Move all images into a folder where they can more easily be replaced
- [ ] Create a sample implementation demonstrating how to white label the app
- [ ] Investigate using xcconfig files to build alternate targets without affecting the core OBA app (i.e. York Regional Transit could have an xcconfig file and a bundle of resources that they either maintain separately or check into this repository that would enable building `com.yorkregionaltransit.onebusaway`, or whatever in a single step)

https://github.com/OneBusAway/onebusaway-android/blob/develop/REBRANDING.md

**Comment by barbeau** [comment]
TIL you can do checkboxes in Github issues.  Thanks @aaronbrethorst! :)

**Comment by barbeau** [comment]
@aaronbrethorst I managed to knock out a fairly good chunk of this on Android (see https://github.com/OneBusAway/onebusaway-android/pull/252).

There are some details we should probably coordinate between iOS and Android, in terms of white-labeling behavior we encourage and set up by default in the samples for each project.  There are obvious ones (e.g., use your own Google Analytics trackers in rebrands), but also ones that are more subtle (e.g., do we allow rebrands to easily opt-out of customer feedback items like email, Twitter URLs and Stop/Trip Problem Reporting, are rebrands assumed to be fixed to a single API URL instead of multi-region, etc.).  Any wisdom/ideas are appreciated, and I'll keep you updated as I refine the Android implementation.

**Comment by aaronbrethorst** [comment]
awesome, thanks sean!

> do we allow rebrands to easily opt-out of customer feedback items like email, Twitter URLs and Stop/Trip Problem Reporting

Seems like unnecessary extra work. If someone wants to submit a PR with that functionality, more power to them—I'd be happy to accept it. Or, if they request the feature and have the time to wait, that's cool too.

> are rebrands assumed to be fixed to a single API URL instead of multi-region, etc.

Again, I'd say whatever the path of least resistance is is probably the best approach for now.

**Comment by bbodenmiller** [comment]
I kind of wonder why it is worth supporting white labeling... seems like if some agency wants white labeling they should come submit the PRs to implement it.

**Comment by aaronbrethorst** [comment]
I'd rather save them the excuse of forking, and it's also a good exercise for me to factor all of our colors, strings and images into a shared location, which is good for my sanity even if no one else ever uses it :)

**Comment by bbodenmiller** [comment]
I agree to a certain extent. Has a specific agency actually expressed interest in releasing OBA under their branding? (In Puget Sound we just spend $80,000 building new apps instead) I'd hope it'd at least say it is "powered by OneBusAway".

In regards to analytics I think it'd be nice to give them to the option to add analytics side by side with the existing tracker rather than replace it. I say this because I think it'd be useful to know how many people are using the OBA code base.

Anyways my primary reason for questioning white labeling revolves around the fact that the most successful transit apps support multiple regions. I wonder if we could seperate branding via the multiregion setup.

**Comment by barbeau** [comment]
>I kind of wonder why it is worth supporting white labeling... seems like if some agency wants white labeling they should come submit the PRs to implement it.

In a perfect world I'd agree, but agencies/companies are currently taking the path of least resistance to get something launched, which means forking the code and changing things, and worrying about maintenance down the road.  To date no one wants to take on the initial overhead of implementing the white-labeling framework - and from an agency's perspective, it can be difficult to justify costs for something that doesn't immediately benefit them (again, easier to fork now and worry about maintenance later).  As @aaronbrethorst says, providing a ready-to-go whitelabeling solution removes any excuse for forking the project, and hopefully concentrates improvements (by everyone) in the same code base.   I also see similar benefits to some project re-org on the Android side of things to update everything to current best practices.

>Has a specific agency actually expressed interest in releasing OBA under their branding?

Yup, see iOS and Android apps in production at [York](http://www.yorkregiontransit.com/en/ridingwithus/apps.asp).  To their credit they released these apps as open-source under Apache as well and as also part of the multi-region solution (although they don't really advertise it - but, check GA and you'll see some usage from that region).  Their tech team saw value in contributing to a white-label framework but wasn't able to sell it up the flagpole.  As mentioned above, removing any additional up-front overhead should hopefully make it much easier to sell the white-label to management.

**Comment by barbeau** [comment]
@aaronbrethorst FYI, whitelabeling is now implemented in the `develop` branch of OBA Android.  The main rebranding readme is [here](https://github.com/OneBusAway/onebusaway-android/blob/develop/REBRANDING.md).

Aside from app id/name and colors, I decided to support two configuration options, modeled after the major modifications that York made to the app:

* Arrival time styling - change the presentation of arrival times to your own style.  We currently support Style A (original OBA presentation) and Style B (York presentation).  Additional styles can be added later.
* Fixed vs. multi-region - York's app was branded only for their agency, and they didn't want it accidentally directing itself to other OBA regions (e.g., upon location error)

More information about configuration options, along with screenshots of the arrival times styles, can be seen [here](https://github.com/OneBusAway/onebusaway-android/blob/develop/REBRANDING.md#configuration-options).

I've also included two sample configurations, "Agency X", and "Agency Y".  "Agency X" is a simple color palette/name swap with OBA, and "Agency Y" has the York modifications.
